### PR TITLE
Add ability to use LoomProvider with the /eth endpoint

### PR DESCRIPTION
**Pull request by redben**

### CODE REVIEW

File: src/helpers.ts [hunk 1]
```diff
@@ -131,7 +131,7 @@ export function setupProtocolsFromEndpoint(
 ): { writer: IJSONRPCClient; reader: IJSONRPCClient } {
   const protocol = selectProtocol(endpoint)
   const writerSuffix = protocol === JSONRPCProtocol.HTTP ? '/rpc' : '/websocket'
-  const readerSuffix = protocol === JSONRPCProtocol.HTTP ? '/query' : '/queryws'
+  const readerSuffix = protocol === JSONRPCProtocol.HTTP ? '/eth' : '/eth'
```
Comment (redben) [inline src/helpers.ts]: Is the ETH endpoint available on http?

File: src/client.ts [hunk 2]
```diff
@@ -81,7 +81,7 @@ export enum ClientEvent {
   /**
    * Exclusively used by loom-provider
    */
-  EVMEvent = 'evmEvent',
+  EVMEvent = 'message', // 'evmEvent',
```
Comment (enlight) [inline src/client.ts]: This looks fishy

File: src/client.ts [hunk 3]
```diff
@@ -646,6 +652,7 @@ export class Client extends EventEmitter {
     hashHexStr: string,
     full: boolean = true
   ): Promise<EthBlockInfo | null> {
+    console.log(hashHexStr)
```
Comment (enlight) [inline src/client.ts]: `console.log` be gone

File: src/client.ts [hunk 4]
```diff
@@ -755,6 +766,12 @@ export class Client extends EventEmitter {
     } else if (result) {
       debugLog('Event', event.id, result)
 
+      // emit evm events as is (jsonrpc/ethereum)
+      if (this.isLegacy === false && isEVM) {
```
Comment (enlight) [inline src/client.ts]: `!this.isLegacy && isEVM`

File: src/helpers.ts (hunk 1 above)
Comment (enlight) [reply src/helpers.ts]: Yes

File: src/internal/ws-rpc-client.ts [hunk 5]
```diff
@@ -35,7 +35,10 @@ export class WSRPCClient extends EventEmitter {
   private _isSubcribed: boolean = false
 
   protected _rpcId: number = 0
-  protected _getNextRequestId = () => (++this._rpcId).toString()
+  // XXX workaround int.string bug on the eth endpoint. Set as string when fixed.
+  // @ts-ignore
+  protected _getNextRequestId = (): string => ++this._rpcId
+  // protected _getNextRequestId = () => (++this._rpcId).toString()
```
Comment (enlight) [inline src/internal/ws-rpc-client.ts]: Switch implementation based on the endpoint.

File: src/loom-provider.ts [hunk 6]
```diff
@@ -158,18 +161,19 @@ export class LoomProvider {
...
+    this._client.readClient.on(eventType, (msg: IChainEventArgs) => this._onWebSocketMessage(msg))
 
     if (!this._setupMiddlewares) {
       this._setupMiddlewares = (client: Client, privateKey: Uint8Array) => {
-        return createDefaultTxMiddleware(client, privateKey as Uint8Array)
+        return createDefaultTxMiddleware(client, privateKey)
       }
     }
+    if (client.isLegacy) this.addLegacyDefaultMethods()
+    else this.addDefaultMethods()
```
Comment (enlight) [inline src/loom-provider.ts]:
```typescript
if (client.isLegacy) {
  this.addLegacyDefaultMethods()
} else {
  this.addDefaultMethods()
}
```

File: src/loom-provider.ts [hunk 7]
```diff
@@ -613,14 +644,52 @@ export class LoomProvider {
...
   }
 
+  private async _ethSubscribe(payload: IEthRPCPayload) {
+    if (!this._subscribed) {
+      this._subscribed = true
+    }
+    const { method, params } = payload // payload.params[0] as string
+
+    const subscriptionId: string = await this._client.evmSubscribeAsync(method, params)
+    // const subscriptionId: string = await this._client.readClient.sendAsync(method, params)
```
Comment (enlight) [inline src/loom-provider.ts]: Remove the commented line

File: src/loom-provider.ts [hunk 8]
```diff
@@ -613,14 +644,52 @@ export class LoomProvider {
...
     return this._client.uninstallEvmFilterAsync(payload.params[0])
   }
 
-  private _ethUnsubscribe(payload: IEthRPCPayload) {
+  private _ethUnsubscribeLegacy(payload: IEthRPCPayload) {
     return this._client.evmUnsubscribeAsync(payload.params[0])
   }
 
+  private async _ethUnsubscribe(payload: IEthRPCPayload) {
+    // return this._client.evmUnsubscribeAsync(payload.params[0])
```
Comment (enlight) [inline src/loom-provider.ts]: Remove line

File: src/client.ts [hunk 9]
```diff
@@ -257,6 +258,10 @@ export class Client extends EventEmitter {
     return this._writeClient.url
   }
 
+  get readClient(): IJSONRPCClient {
+    return this._readClient
+  }
```
Comment (enlight) [inline src/client.ts]: Doesn't seem to be used anywhere.

File: src/internal/ws-rpc-client.ts [hunk 10]
```diff
@@ -211,11 +214,14 @@ export class WSRPCClient extends EventEmitter {
...
-
-    // Events from EVM have the id from the evmsubscribe command
-    if (/^0x.+$/.test(msg.id)) {
-      log('EVM Event arrived', msg)
+    // EVM Event old endpoint
+    else if (/^0x.+$/.test(msg.id)) {
       this.emit(RPCClientEvent.EVMMessage, this.url, msg)
     }
+    // EVM Event eth endpoint
+    else if (msg.params && 'subscription' in msg.params) {
```
Comment (enlight) [inline src/internal/ws-rpc-client.ts]: Don't quite follow this

File: src/client.ts [hunk 11]
```diff
@@ -547,15 +554,14 @@ export class Client extends EventEmitter {
...
     debugLog(`Get filter changes for ${JSON.stringify({ id }, null, 2)}`)
-    const result = await this._readClient.sendAsync<Uint8Array>('getevmfilterchanges', {
+    const result = await this._readClient.sendAsync<string>('getevmfilterchanges', {
       id
     })
 
     if (result) {
       const envelope: EthFilterEnvelope = EthFilterEnvelope.deserializeBinary(
-        bufferToProtobufBytes(result)
+        B64ToUint8Array(result)
```
Comment (enlight) [inline src/client.ts]: Why did this method change at all?

File: src/tests/e2e/loom-provider-web3-tests.ts [hunk 12]
```diff
@@ -321,7 +319,8 @@ test('LoomProvider + Web3 + Logs', async t => {
     })
     console.log('events', events)
     t.assert(events.length > 0, 'Should have more than 0 events')
-    t.assert(events[0].blockTime > 0, 'blockTime should be greater than 0')
+    // no blockTime in events
+    // t.assert(events[0].blockTime > 0, 'blockTime should be greater than 0')
```
[the remark this answers is not in the document]
Comment (redben) [reply src/tests/e2e/loom-provider-web3-tests.ts]: Will we have this extension on the new eth endpoint later?

File: src/tests/e2e/loom-provider-web3-tests.ts [hunk 13]
```diff
@@ -236,14 +236,10 @@ test('LoomProvider + Web3 + getBlockByNumber', async t => {
...
-    const blockInfo = await web3.eth.getBlock(blockNumber, false)
-    const blockInfoByHash = await web3.eth.getBlock(blockInfo.transactionHash, false)
-    t.assert(blockInfoByHash, 'Should return block info by hash')
-  } catch (err) {
-    console.log(err)
-  }
+  const blockNumber = await web3.eth.getBlockNumber()
+  const blockInfo = await web3.eth.getBlock(blockNumber, false)
+  const blockInfoByHash = await web3.eth.getBlock(blockInfo.hash, false)
+  t.assert(blockInfoByHash, 'Should return block info by hash')
```
[the remark this answers is not in the document]
Comment (redben) [reply src/tests/e2e/loom-provider-web3-tests.ts]: What was it for?

File: src/tests/e2e/loom-provider-web3-tests.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (enlight) [reply src/tests/e2e/loom-provider-web3-tests.ts]: Yeah, we'll probably add it back in.

File: src/helpers.ts (hunk 1 above)
Comment (enlight) [inline src/helpers.ts]: -> `const readerSuffix = '/eth'`

File: src/tests/e2e/loom-provider-web3-tests.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (enlight) [reply src/tests/e2e/loom-provider-web3-tests.ts]: Looks like it was logging the error

File: src/tests/e2e/loom-provider-web3-tests.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (redben) [reply src/tests/e2e/loom-provider-web3-tests.ts]: Ok. Added a comment to reenable once that's done.

File: src/tests/e2e/loom-provider-web3-tests.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (redben) [reply src/tests/e2e/loom-provider-web3-tests.ts]: Reverted.

File: src/client.ts (hunk 2 above)
Comment (redben) [reply src/client.ts]: Reverted.

File: src/helpers.ts (hunk 1 above)
Comment (redben) [reply src/helpers.ts]: Done.

**Comment by redben** [comment]
One test still fails...

**Calls via HTTP**
```
contractAddr = await client.getContractAddressAsync('BluePrint')
>> Error: Failed to resolve contract address
 ```

**Comment by enlight** [comment]
This is an intermediate step that will allow us to start using the Web3 JSON-RPC methods provided by Loom nodes on the `/eth` endpoint. The end goal is to switch over to a completely new version of `LoomProvider` that's being built in https://github.com/loomnetwork/loom-js/pull/303 but that still requires some work in loomchain, as well as extensive testing.

One annoying limitation of the changes in this PR is that a single `Client` instance can't be used to interact with both the `/eth` and `/query` endpoints, so any one `Client` instance can only interact with EVM contracts, or Go contracts, but not both at once.

**Pull request merged**